dtfs-central-api/src/types/utilisation-reports.ts

```diff
@@ -15,11 +16,19 @@ export type GetUtilisationReportResponse = {
   id: number;
   bankId: string;
   status: UtilisationReportReconciliationStatus;
-  uploadedByUserId: string | null;
   reportPeriod: ReportPeriod;
-  azureFileInfo: AzureFileInfo | null;
-  dateUploaded: Date | null;
-};
+} & (
+  | {
+      uploadedByUser: UploadedByUserDetails;
+      azureFileInfo: AzureFileInfo;
+      dateUploaded: Date;
+    }
+  | {
+      uploadedByUser: null;
+      azureFileInfo: null;
+      dateUploaded: null;
+    }
+);
```

Comment:
This feels like a good use-case for a more narrow type as our mapping function used by the controllers can ensure that we are either in an "uploaded" or "not uploaded" state

Reply:
Oh nice, I didn't consider this when I wrote it but perhaps should have done! That said, whilst I agree this typing portrays the returned object better and we should keep it now you have gone to the effort of doing it, I do just want to highlight how I see some of the trade offs, to ensure we weigh up the benefits against the dev effort when doing refactors of this kind in the future:

- With this change, I would say we now *have* to assert it falls into one of the two types when mapping to this, when we already know it will be one of the two because we wrote the code that gets reports into these states, except now we have more code to maintain. (and actually because here this is just being returned as the response immediately after being mapped to, and we already knew it would be one of the two shapes, we could have left it and still safely parse the response in say portal api with the stricter type without bothering to do so here (although I will accept that could look confusing))
- Narrowing types can sometimes be good for documentation as code. In this specific case, I think it was already fairly clear which fields would and would not be populated when the file is or isn't uploaded, but this does make it concrete for anyone less familiar with the domain!
- Anywhere objects are of this type, if we want to access azureFileInfo, will still either have to check it has a value or not, or assert which of the two possible types the main object falls into, which is not necessarily simpler. This can have a lot of value if we are filtering to one of the subtypes and then we don't have to be continually checking for field existence before using it, but I'm not sure we have any cases like that? (e.g. if we had left it we could still introduce a tighter type in say portal if we ever needed to to benefit from knowledge of their combined existence)

That all said, let's keep it! But, I do wonder if we can make the type slightly more readable by maybe extracting the types we are doing the operations on? One option which is perhaps a bit lengthy but prevent us having to define weird partial types that don't have individual domain meaning would be to have the underlying type that is as was before but not exportable, and then have a derived type where the fields are required and another where they are null, and then our endpoint actually just returns one | the other? What do you think?